### PR TITLE
Persistence

### PR DESCRIPTION
**Pull request by lock9**
Basic persistence documentation

### CODE REVIEW

File: en/6-persistence/README.md [hunk 1]
```diff
@@ -1,4 +1,503 @@
...
-    -  Level DB
-    -  The table structure in the Level DB
-    -  wallet index
\ No newline at end of file
+
+# Persistence
+
+## Serialization Model
+The native types in NEO platform are Integer, Decimal and String, UInt160 and UInt256.
+This model is used for both the persistence and the network layer.
```
Comment (Tommo-L) [inline en/6-persistence/README.md]: test

File: en/6-persistence/README2.md [hunk 2]
```diff
@@ -0,0 +1,507 @@
+### LevelDB Blockchain Data Structure
+The current implementation of NEO in C# uses LevelDB, a fast key-value database, to persist blockchain information. This DB is used for both system data, like blocks and transactions, but also for smart contract data.  
+The data is structured using *prefixes*.
+We can see here the prefixes used in our C# version:
+
+
+| Prefix     | Data     |
```
Comment (Tommo-L) [inline en/6-persistence/README2.md]: 1. Could add `LevelDB Key`  and `LevelDB Value` , `description` here?
2. Be careful with the `LevelDB Value` of the `Block` and the `LevelDB Key` of the `Smart Contract Storage`  table. The `LevelDB Value` of the `Block`  table is `the sum of system fee  + block data`.  The `LevelDB Key` of the `Smart Contract Storage` is `Prefix + the script hash of the contract + Key`.
3. Maybe the `Block headers` change to `Block` will be better, as block headers and blocks are the same table.

@steven1227 @lock9

File: en/6-persistence/README2.md [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+The memory pool is a collection in memory used to track transactions that are not yet committed/persisted.
+When looking for a transaction, the node will always check first if the transaction is in the memory pool, before checking the device storage.
+
+![memory_pool_1](persistence_memory_pool_1.png)
+
+If the `Storage`  returns a null transaction, either the transaction does not exist (if the node is fully synchronized), or the transaction must be retrieved from other nodes.
+
+
+#### 0x01 Block Headers (blocks)
+This is where all block information is stored without it's transactions. A block is the block-header with transactions, so even if a block hash is found in this collection, it doesn't mean the block is fully synchronized. It can only be considered a block after it has transactions added to it.
```
Comment (Tommo-L) [inline en/6-persistence/README2.md]: Could add  the structure of each table and one or two usage case, including the `blocks`, `transactions,` `accounts`,` coins`, `spent coins`, `validators`, `assets`, `contracts`, `storages`, etc.

Comment (Tommo-L) [reply en/6-persistence/README2.md]: @steven1227 @lock9

File: en/6-persistence/README2.md [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+          }
+      }
+      return true;
+  }
+
+```
+
+
+#### 0x40 - Accounts
+
```
Comment (Tommo-L) [inline en/6-persistence/README2.md:141]: The `Accounts` table storage the user information, votes, and the balance of global assets and NEP-5 assets.
For more details, you can see https://github.com/neo-project/neo/tree/master-2.x/neo/Ledger
@steven1227 @lock9

File: en/6-persistence/README2.md [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+      batch.Delete(it.Key());
+    }
+  }
+  db.Put(WriteOptions.Default, SliceBuilder.Begin(Prefixes.SYS_Version), Assembly.GetExecutingAssembly().GetName().Version.ToString());
+  db.Write(WriteOptions.Default, batch);
+}
+```
+
+Wallet usage:
+
```
Comment (Tommo-L) [inline en/6-persistence/README2.md:482]: How about removing this subchapter `wallet index`? As the wallet index is used in CLI and GUI wallets, is not the core      function.  @steven1227 @lock9

File: en/6-persistence/README.md [hunk 6]
```diff
@@ -1,4 +1,503 @@
...
+public byte[] Hash160(byte[] message)
+{
+  return message.Sha256().RIPEMD160();
+}
+```
+
+#### Address example
+UInt160 is used for contract hashes and the resulting hash is used to build the address.  
+The address is the script-hash of the contract, encoded in Bas58 prefixed with the contract version, currently `0x23`.
+
```
Comment (Tommo-L) [inline en/6-persistence/README.md:152]: `0x23` should be change to `0x17`,  `0x17` = 23

Comment (Tommo-L) [reply en/6-persistence/README.md:152]: @steven1227 @lock9

Comment (Tommo-L) [reply en/6-persistence/README.md:152]: `Bas58` should be changed to `Base58Check`

**Comment by lock9** [comment]
Hello!
Changes requested:
- Add LevelDB Keys / Description in the table;
- "Be careful with the LevelDB Value of the Block and the LevelDB Key of the Smart Contract Storage table. The LevelDB Value of the Block table is the sum of system fee + block data. The LevelDB Key of the Smart Contract Storage is Prefix + the script hash of the contract + Key." - I didn't know that!   Thanks for noticing.
- Good suggestion about the block headers and blocks, I agree that we can use only "blocks" and "forget" about block headers. 
-  Add examples of usage of each table `blocks`, `transactions`, `accounts`, `coins`, `spent coins`, `validators`, `assets`, `contracts`, `storages` (and other). I'm thinking about adding UML diagrams and their corresponding code. Is it ok for you? 
- Increment the `accounts` table description. I'm not very familiar with that, I know that they store the user balance, but I don't actually understand why. I never understood why store this value if we depend on UTXO. I'm going to use the link you sent. Thanks!
- Remove the wallet part. I agree. 

Is there anything else you missed and you think we should add?

**Comment by Tommo-L** [comment]
> Add examples of usage of each table blocks, transactions, accounts, coins, spent coins, validators, assets, contracts, storages (and other). I'm thinking about adding UML diagrams and their corresponding code. Is it ok for you?

I think its enough.

> Increment the accounts table description. I'm not very familiar with that, I know that they store the user balance, but I don't actually understand why. I never understood why store this value if we depend on UTXO.

It's mainly used in two cases:
1. queries in RPC, see this link: https://github.com/neo-project/neo/blob/master-2.x/neo/Network/RPC/RpcServer.cs#L401 
2. the votes for validators. see this link: https://github.com/neo-project/neo/blob/master-2.x/neo/Ledger/Blockchain.cs#L500 

> Is there anything else you missed and you think we should add?

No

**Comment by lock9** [comment]
@Tommo-L @steven1227 
Hello, I've made some changes as requested, except for the examples.
Could you elucidate more what do you expect in the examples? 
I can see what messages trigger a state change, however, I'm not sure "how to add it as an example". 

I can add a code snippet, but I don't know how to "introduce the example". Any suggestions?

File: en/6-persistence/README2.md [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+| 0x48      | **Validators**: Network Validators meta-data        | Validator Public Key | [ValidatorState](https://github.com/neo-project/neo/blob/master-2.x/neo/Ledger/ValidatorState.cs)|
+| 0x4c      | **Assets**: Native coins meta-data       | Register Transaction Hash | [AssetState](https://github.com/neo-project/neo/blob/master-2.x/neo/Ledger/AssetState.cs) |
+| 0x50      | **Contracts**: Contract meta-data       | Contract Script Hash| [ContractState](https://github.com/neo-project/neo/blob/master-2.x/neo/Ledger/ContractState.cs) |
+| 0x70    | **Smart Contract Storage**: Storage used by SmartContracts       | ScriptHash + Key | Byte[] |
+| 0x80      | **Pending Headers**: Used in synchronization to track blocks waiting to be synced        | Prefix only | [HeaderHashList](https://github.com/neo-project/neo/blob/master-2.x/neo/Ledger/HeaderHashList.cs) |
+| 0x90      | **Validators Count**: TODO         | Prefix only| [ValidatorsCountState](https://github.com/neo-project/neo/blob/master-2.x/neo/Ledger/ValidatorsCountState.cs) |
```
Comment (Tommo-L) [inline en/6-persistence/README2.md]: `Validators Count` is used for recording the votes of  how many consensus nodes to be selected.  More detail about voting: https://docs.neo.org/developerguide/en/articles/consensus/vote_validator.html

File: en/6-persistence/README2.md [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+```CSharp
+UInt160[] IVerifiable.GetScriptHashesForVerifying(Snapshot snapshot)
+{
+  ECPoint[] validators = snapshot.GetValidators();
+  if (validators.Length <= ValidatorIndex)
+  throw new InvalidOperationException();
+  return new[] { Contract.CreateSignatureRedeemScript(validators[ValidatorIndex]).ToScriptHash() };
+}
+```
+
```
Comment (Tommo-L) [inline en/6-persistence/README2.md:232]: Maybe, we replace by this code is better: https://github.com/neo-project/neo/blob/master-2.x/neo/Consensus/ConsensusContext.cs#L395
As `Validators` is mainly for electing the consensus nodes.  The `NextConsensus` field in block, points to the consesus nodes in the next round.  more details: https://docs.neo.org/developerguide/en/articles/consensus/vote_validator.html

**Comment by Tommo-L** [comment]
> @Tommo-L @steven1227
> Hello, I've made some changes as requested, except for the examples.
> Could you elucidate more what do you expect in the examples?
> I can see what messages trigger a state change, however, I'm not sure "how to add it as an example".
> 
> I can add a code snippet, but I don't know how to "introduce the example". Any suggestions?


I think it is not necessary to add more example, just add code description is enough. As you have added `invoking` code bellow those collections.

**Comment by lock9** [comment]
@Tommo-L  changes added

**Comment by Tommo-L** [comment]
thanks

**Comment by Tommo-L** [comment]
@steven1227  We could merge now.

**Pull request merged**